src/smirnoffio.py

```diff
@@ -629,6 +613,53 @@ def build_system_mval_masks(self):
         logger.info("system_mval_masks is built for faster gradient evaluations")
         self.system_mval_masks = system_mval_masks
 
+class TorsionProfileTarget_SMIRNOFF(TorsionProfileTarget):
+    """ Force and energy matching using SMIRKS native Open Force Field (SMIRNOFF). """
+    def __init__(self,options,tgt_opts,forcefield):
+        ## Default file names for coordinates and key file.
+        self.set_option(tgt_opts,'pdb',default="conf.pdb")
+        # List of .mol2 files for SMIRNOFF to set up the system
+        self.set_option(tgt_opts,'mol2',forceprint=True)
+        self.set_option(tgt_opts,'coords',default="scan.xyz")
+        self.set_option(tgt_opts,'openmm_precision','precision',default="double", forceprint=True)
+        self.set_option(tgt_opts,'openmm_platform','platname',default="Reference", forceprint=True)
+        self.engine_ = SMIRNOFF
+        ## Initialize base class.
+        super(TorsionProfileTarget_SMIRNOFF,self).__init__(options,tgt_opts,forcefield)
+
+    def submit_jobs(self, mvals, AGrad=False, AHess=False):
+        # we update the self.pgrads here so it's not overwritten in rtarget.py
+        self.smirnoff_update_pgrads()
+
+    def smirnoff_update_pgrads(self):
+        """
+        Update self.pgrads based on smirks present in mol2 files
+
+        This can greatly improve gradients evaluation in big optimizations
+
+        Note
+        ----
+        1. This function assumes the names of the forcefield parameters has the smirks as the last item
+        2. This function assumes params only affect the smirks of its own. This might not be true if parameter_eval is used.
+        """
+        orig_pgrad_set = set(self.pgrad)
+        # smirks to param_idxs map
+        smirks_params_map = defaultdict(list)
+        for pname, pidx in self.FF.map.items():
```

Comment:
This is the old version to determine the mapping between smirks and affected param_idxs, I think we should use the updated version similar to other targets, like this:
```
        # New code for mapping smirks to mathematical parameter IDs
        for pname in self.FF.pTree:
            smirks = pname.rsplit('/',maxsplit=1)[-1]
            # print("pname %s mathid %s -> smirks %s" % (pname, str(self.FF.get_mathid(pname)), smirks))
            for pidx in self.FF.get_mathid(pname):
                smirks_params_map[smirks].append(pidx)
```

Reply:
Changes have been made in the latest commit.